### PR TITLE
SE-2176 Update doc comments for accuracy

### PR DESCRIPTION
**Pull request by swalladge**
Currently there are discrepancies between the official docs, the code, and the doc comments here. These were found while investigating https://github.com/edx/edx-platform/pull/23039#discussion_r417670046

This PR does:
- provide example of html escaping for the interpolateHtml method
- fix misleading comments about escaping for StringUtils.interpolate (it
  does not in fact escape anything)
- update interpolate examples for accuracy and to better represent functionality
- add a test to show that html is not escaped

**Jira tickets**: [OSPR-4435](https://openedx.atlassian.net/browse/OSPR-4435)

**Test instructions**:

- cross reference the updated doc comments with the code, tests, and [official docs](https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/preventing_xss/preventing_xss.html#legacy-javascript-files) to verify accuracy
- verify that tests pass

**Reviewers**:

- [x] @bradenmacdonald 
- [ ] edX reviewers TBD

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @swalladge! I've created [OSPR-4435](https://openedx.atlassian.net/browse/OSPR-4435) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

**Comment by natabene** [comment]
@swalladge Thank you for your contribution. Please let me know once it is ready for our review.

File: src/js/utils/string-utils.js [hunk 1]
```diff
@@ -16,19 +16,18 @@
...
          *
          *~~~ javascript
-         * HtmlUtils.interpolate(
+         * StringUtils.interpolate(
          *     'You are enrolling in {spanStart}{courseName}{spanEnd}',
          *     {
          *         courseName: 'Rock & Roll 101',
-         *         spanStart: HtmlUtils.HTML('<span class="course-title">'),
+         *         spanStart: '<span class="course-title">',
          *         spanEnd: HtmlUtils.HTML('</span>')
```
Comment (bradenmacdonald) [inline src/js/utils/string-utils.js:31]: I agree that your change here matches the [docs](https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/preventing_xss/preventing_xss.html#legacy-javascript-files), but this is a bad example. I guess someone copy-pasted the docs from `HtmlUtils.interpolate` to become the docs to `StringUtils.interpolate` and didn't update it.

If I'm understanding this correctly: In the case of `StringUtils.interpolate`, it wouldn't make any sense to deliberately pass in some `HtmlUtils.HTML` as parameters, because the output of `StringUtils.interpolate` is a plain string, and the `HtmlUtils.HTML` safety of the HTML will get lost, resulting in that HTML being escaped anyways.

The example output below, `'You are enrolling in <span class="course-title">Rock & Roll 101</span>'`, while correct, seems to imply that the HTML is desired and that the whole string would somehow be passed to an end-user as-is, which is a security issue because it contains unescaped user input. Or, if this was properly escaped later ("delaying escaping as late as possible"), the result that users see would be `You are enrolling in &lt;span class="course-title"&gt;Rock &amp; Roll 101&lt;/span&gt;` which seems clearly wrong.

A better example would just be

```
StringUtils.interpolate(
    'You are enrolling in {courseName}',
    {courseName: 'Rock & Roll 101'},
);
```

with the output `"You are enrolling in Rock & Roll 101"; // plain string, unescaped`

File: src/js/utils/html-utils.js [hunk 2]
```diff
@@ -86,17 +86,17 @@
          * HtmlUtils.interpolateHtml(
          *     'You are enrolling in {spanStart}{courseName}{spanEnd}',
          *     {
-         *         courseName: 'Rock & Roll 101',
+         *         courseName: '<b>Rock</b> & Roll 101',
```
Comment (robrap) [inline src/js/utils/html-utils.js:89]: I don't think adding the `<b>` really helps here, because it would look like a bug in the final HTML.

File: src/js/utils/html-utils.js [hunk 3]
```diff
@@ -86,17 +86,17 @@
...
          *     }
          * );
          *~~~
          *
-         * returns:
+         * returns an HtmlSnippet object whose .toString() method returns:
          *
          *~~~ javascript
-         * 'You are enrolling in <span class="course-title">Rock &amp; Roll 101</span>'
+         * 'You are enrolling in <span class="course-title">&lt;b&gt;Rock&lt;/b&gt; &amp; Roll 101</span>'
```
Comment (robrap) [inline src/js/utils/html-utils.js:99]: I think I'd revert this change based on my earlier comment.

File: src/js/utils/string-utils.js [hunk 4]
```diff
@@ -16,19 +16,18 @@
          * indicated via curly braces, e.g. 'Hello {name}'. These tokens are
          * replaced by the parameter value of the same name.
          *
-         * Parameter values will be rendered using their toString methods and then
-         * HTML-escaped. The only exception is that instances of the class HTML
-         * are rendered without escaping as their contract declares that they are
-         * already valid HTML.
+         * Parameter values will be rendered using their toString methods.
+         * **NO** HTML escaping or sanitizing of any form is performed.
+         * If HTML escaping is required, use HtmlUtils.interpolateHtml().
```
Comment (robrap) [inline src/js/utils/string-utils.js:21]: 👍

File: src/js/utils/string-utils.js (hunk 1 above)
Comment (swalladge) [reply src/js/utils/string-utils.js:31]: @bradenmacdonald @robrap yep fair enough, I see what you're saying. My intention for showing those in the examples was to be absolutely clear what does html escaping and what does not. I'll change them though

File: src/js/utils/html-utils.js (hunk 2 above)
Comment (swalladge) [reply src/js/utils/html-utils.js:89]: Do you have a suggestion for an example showing html actually being escaped?

File: src/js/utils/string-utils.js (hunk 1 above)
Comment (swalladge) [reply src/js/utils/string-utils.js:31]: @bradenmacdonald also your example isn't a great real-world example, because no-one should actually do that with escaping `courseName`.

Comment (swalladge) [reply src/js/utils/string-utils.js]: @robrap  @bradenmacdonald please see the changes I made - I followed your suggestions, and added more examples for clarity.

Comment (bradenmacdonald) [reply src/js/utils/string-utils.js]: 👍 LGTM.

File: src/js/utils/html-utils.js (hunk 2 above)
Comment (robrap) [reply src/js/utils/html-utils.js]: I was going to say the `&` is escaped.  Maybe that is why you resolved this?

File: src/js/utils/string-utils.js [hunk 5]
```diff
@@ -16,37 +16,36 @@
...
+         *         n: comments.length,
          *     }
          * );
          *~~~
          *
          * returns:
          *
          *~~~ javascript
-         * 'You are enrolling in <span class="course-title">Rock &amp; Roll 101</span>'
+         * '2 comments'
```
Comment (robrap) [inline src/js/utils/string-utils.js:39]: I think that `'You are enrolling in {courseName}'`, where `courseName: 'Rock & Roll 101',`, would be a good example that shows that `&` is not escaped in the output.

File: src/js/utils/html-utils.js (hunk 2 above)
Comment (swalladge) [reply src/js/utils/html-utils.js]: I had thought of another example, but yep what you suggested is good. :+1:

**Comment by bradenmacdonald** [comment]
👍 LGTM FWIW though I'm not super familiar with this area.

File: src/js/utils/string-utils.js [hunk 6]
```diff
@@ -16,37 +16,35 @@
...
          *
          * Note: typically the formatString will need to be internationalized, in which
          * case it will be wrapped with a call to an i18n lookup function. In Django,
          * this would look like:
          *
          *~~~ javascript
-         * HtmlUtils.interpolate(
-         *     gettext('You are enrolling in {spanStart}{courseName}{spanEnd}'),
+         * StringUtils.interpolate(
+         *     gettext('{n} comments'),
```
Comment (robrap) [inline src/js/utils/string-utils.js:47]: Maybe update this to `gettext('You are enrolling in {courseName}')`?  Fewer changes from the above example will call attention to what is actually different here.

Comment (swalladge) [reply src/js/utils/string-utils.js]: Oops missed that one. :+1:

**Comment by natabene** [comment]
@robrap Can you give another look when you have a chance?

File: src/js/utils/specs/string-utils-spec.js [hunk 7]
```diff
@@ -27,6 +27,10 @@ define(
                         'Hello, {name}. Here is a { followed by a }', {name: 'Andy'},
                         'Hello, Andy. Here is a { followed by a }'
                     ]
```
Comment (robrap) [inline src/js/utils/specs/string-utils-spec.js:29]: Missing a comma.
Maybe this is breaking the build? The error reported in Travis is not helpful.  I re-ran master and it is passing, so hopefully this is it.

Comment (robrap) [reply src/js/utils/specs/string-utils-spec.js:29]: swalladge: Maybe you missed the first line above?  There is a bug.  You are missing a comma at the end of line 29.

Comment (swalladge) [reply src/js/utils/specs/string-utils-spec.js]: Thanks, nice catch. 

The unhelpful travis error is due to recent changes with how xvfb works with travis (the previous builds including the one you reran were on trusty, travis now defaults to xenial). I'm going to try using the latest method for xvfb with xenial, otherwise will explicitly pin back to trusty and use the original method.

Comment (swalladge) [reply src/js/utils/specs/string-utils-spec.js]: Tests passing now, i'll squash in prep for merge. Thanks for your review @robrap ! :)

**Comment by edx-webhook** [comment]
@swalladge 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-ui-toolkit&entry.867055334=https://github.com/edx/edx-ui-toolkit/pull/109&entry.1484655318=https://github.com/swalladge&entry.752974735=2020-04-30+00:49&entry.1917517419=2020-05-08+00:15&entry.2133058324=Yes) so we can improve your experience in the future.
